### PR TITLE
Resolve deferred callable before passing it to foundHandler for invocation

### PR DESCRIPTION
**Pull request by astax-t**
Fixes issue #1785

Includes https://github.com/slimphp/Slim/pull/1783

### CODE REVIEW

File: Slim/DeferredCallable.php [hunk 1]
```diff
@@ -27,13 +35,15 @@ public function __construct($callable, ContainerInterface $container = null)
 
     public function __invoke()
     {
+        return call_user_func_array($this->getResolvedCallable(), func_get_args());
+    }
+
+    public function getResolvedCallable()
+    {
         $callable = $this->resolveCallable($this->callable);
```
Comment (akrabat) [inline Slim/DeferredCallable.php:43]: Should we do this only once? i.e. should `getResolvedCallable()` store the resolved callable into `$this->resolvedCallable` and then return that rather than re-resolving?

File: Slim/Route.php [hunk 2]
```diff
@@ -314,13 +314,19 @@ public function __invoke(ServerRequestInterface $request, ResponseInterface $res
         /** @var InvocationStrategyInterface $handler */
         $handler = isset($this->container) ? $this->container->get('foundHandler') : new RequestResponse();
 
+        if ($this->callable instanceof DeferredCallable) {
+            $callable = $this->callable->getResolvedCallable();
+        } else {
+            $callable = $this->callable;
+        }
```
Comment (akrabat) [inline Slim/Route.php:321]: Why is this required? i.e. if the problem is that you can't get at the real callable in the invocation strategy, then this is solved by the availability of `getResolvedCallable()` which could be called in the the invocation strategy's `__invoke()` if required?

**Comment by akrabat** [comment]
I think #1789 is the correct way to solve this.

File: Slim/Route.php (hunk 2 above)
Comment (astax-t) [inline Slim/Route.php:321]: It's possible, of course, but `DeferredCallable` is an implementation detail and my custom `InvocationStrategy` doesn't need to know about it. So it's reasonable to expect it to get the original callable as the first argument but not some wrapper.

Alternatively, this should be documented.

Comment (akrabat) [inline Slim/Route.php:321]: You are correct, which is why #1789 is a nicer solution as we don't actually need a `DeferredCallable` at all for the route callable.

**Comment by astax-t** [comment]
I agree, https://github.com/slimphp/Slim/pull/1789 is better way
